misc/test_runner/test_runner.c

```diff
@@ -97,31 +97,55 @@ int main() {
     tests[i] = *test_info;
   }
 
+  // Check test values and merge them (e.x. `signal` and `exit_code`).
   for (i = 0; i < num_tests; i++) {
-    const struct fake_criterion_test *test_info = &tests[i];
-    /* Run a single test without checking exit status if IA2_TEST_NAME is set. */
-    bool single_test = getenv("IA2_TEST_NAME") != NULL;
-    if (single_test) {
-      if (strcmp(test_info->name, getenv("IA2_TEST_NAME")))
-        continue;
-      fprintf(stderr, "running single test alone (exit status will not be checked):\n");
+    struct fake_criterion_test *test = &tests[i];
+    if (test->signal != 0) {
+      // If an expected signal was set, check if it was valid.
+      assert(test->signal > 0 && test->signal < _NSIG);
+      if (test->exit_code != 0) {
+        fprintf(stderr, "can't expect both signal %d (SIG%s) and exit status %d\n",
+                test->signal, sigabbrev_np(test->signal), test->exit_code);
+        return EXIT_FAILURE;
+      }
+      // Sometimes the signal doesn't show up with `WIFSIGNALED`,
+      // so we check the exit status as well.
+      test->exit_code = 128 + test->signal;
     }
+    if (test->exit_code > 128 && test->exit_code < 128 + _NSIG) {
+      test->signal = test->exit_code - 128;
+    }
+  }
+
+  // Run a single test without checking exit status if `IA2_TEST_NAME` is set.
+  const char *const single_test_name = getenv("IA2_TEST_NAME");
 
-    const int exit_code = test_info->exit_code;
-    fprintf(stderr, "running suite '%s' test '%s', expecting exit code %d",
-            test_info->suite, test_info->name, exit_code);
-    if (exit_code == EXIT_SUCCESS) {
-      fprintf(stderr, " (%s)", STRINGIFY(EXIT_SUCCESS));
-    } else if (exit_code == EXIT_FAILURE) {
-      fprintf(stderr, " (%s)", STRINGIFY(EXIT_FAILURE));
-    } else if (exit_code > 128) {
-      fprintf(stderr, " (SIG%s)", sigabbrev_np(exit_code - 128));
+  for (i = 0; i < num_tests; i++) {
+    const struct fake_criterion_test *test = &tests[i];
+
+    fprintf(stderr, "running suite '%s' test '%s', expecting ", test->suite, test->name);
+    if (test->signal != 0) {
+      fprintf(stderr, "signal %d (SIG%s)", test->signal, sigabbrev_np(test->signal));
+    } else {
+      fprintf(stderr, "exit status %d", test->exit_code);
+      if (test->exit_code == EXIT_SUCCESS) {
+        fprintf(stderr, " (%s)", STRINGIFY(EXIT_SUCCESS));
+      } else if (test->exit_code == EXIT_FAILURE) {
+        fprintf(stderr, " (%s)", STRINGIFY(EXIT_FAILURE));
+      }
     }
     fprintf(stderr, "...\n");
 
+    if (single_test_name) {
+      if (strcmp(test->name, single_test_name) != 0) {
+        continue;
+      }
+      fprintf(stderr, "running single test alone (exit status will not be checked):\n");
+    }
+
```

Comment:
These two sections should have their order swapped. Otherwise when `IA2_TEST_NAME` is set the program misleadingly claims that it is running other tests in the suite.

Reply:
Fixed in f65be24.